spec/test.spec.js

```diff
@@ -279,60 +279,141 @@ describe('S3Adapter tests', () => {
+        upload: (params, callback) => setTimeout(() => {
+          const { Key, Body } = params;
+
+          objects[Key] = Body;
 
-        if (objects[Key]) {
           callback(null, {
-            Body: Buffer.from(objects[Key], 'utf8'),
+            Location: `https://${bucket}.s3.amazonaws.com/${Key}`,
           });
-        } else {
-          callback(new Error('Not found'));
-        }
-      }, 100),
-    };
+        }, 100),
+        deleteObject: (params, callback) => setTimeout(() => {
+          const { Key } = params;
+
+          delete objects[Key];
+
+          callback(null, {});
+        }, 100),
+        getObject: (params, callback) => setTimeout(() => {
+          const { Key } = params;
+
+          if (objects[Key]) {
+            callback(null, {
+              Body: Buffer.from(objects[Key], 'utf8'),
+            });
+          } else {
+            callback(new Error('Not found'));
+          }
+        }, 100),
+      };
+    }
+    return s3;
   }
 
-  filesAdapterTests.testAdapter('S3Adapter', s3);
+  describe('preserveFilename', () => {
+    let options;
+    const promises = [];
+
+    beforeEach(() => {
+      options = {
+        fileNameCheck: 'loose',
+        preserveFileName: 'never',
+      };
+    });
+
+    it('should add a unique timestamp to the file name when the preserveFileName option is never', () => {
+      options.preserveFileName = 'never';
+      const s3 = makeS3Adaptor(options);
+      const fileName = 'randomFileName.txt';
+      const response = s3.createFile(fileName, 'hello world', 'text/utf8').then((value) => {
+        const url = new URL(value.Location);
+        expect(url.pathname.indexOf(fileName) > 13).toBe(true);
+      });
+      promises.push(response);
+    });
+
+    it('should not add unique timestamp to the file name when the preserveFileName option is hasPath and there is a path', () => {
+      options.preserveFileName = 'hasPath';
+      const s3 = makeS3Adaptor(options);
+      const fileName = 'foo/randomFileName.txt';
+      const response = s3.createFile(fileName, 'hello world', 'text/utf8').then((value) => {
+        const url = new URL(value.Location);
+        expect(url.pathname.substring(1)).toEqual(fileName);
+      });
+      promises.push(response);
+    });
+
+    it('should add unique timestamp to the file name after the last directory when the preserveFileName option is never and there is a path', () => {
+      options.preserveFileName = 'never';
+      const s3 = makeS3Adaptor(options);
+      const fileName = 'foo/randomFileName.txt';
+      const response = s3.createFile(fileName, 'hello world', 'text/utf8').then((value) => {
+        const url = new URL(value.Location);
+        expect(url.pathname.indexOf('foo/')).toEqual(1);
+        expect(url.pathname.indexOf('random') > 13).toBe(true);
+      });
+      promises.push(response);
+    });
+
+    afterAll(() => Promise.all(promises));
```

Comment:
Don't forget to do this.   Otherwise, they resolve in filesAdapterTests.testAdapter(), which breaks reporting.